internal/consensus/reactor.go

```diff
@@ -1003,6 +778,198 @@ OUTER_LOOP:
 	}
 }
 
+// pick a block part to send if the peer has the same part set header as us or if they're catching up and we have the block.
+// returns the part and a bool that signals whether to continue to the loop (true) or to sleep.
+// NOTE there is one case where we don't return a part but continue the loop (ie. we return (nil, true)).
+func pickPartToSend(
+	logger log.Logger,
+	blockStore sm.BlockStore,
+	rs *cstypes.RoundState,
+	ps *PeerState,
+	prs *cstypes.PeerRoundState,
+) (*types.Part, bool) {
+	// If peer has same part set header as us, send block parts
+	if rs.ProposalBlockParts.HasHeader(prs.ProposalBlockPartSetHeader) {
+		if index, ok := rs.ProposalBlockParts.BitArray().Sub(prs.ProposalBlockParts.Copy()).PickRandom(); ok {
+			part := rs.ProposalBlockParts.GetPart(index)
+			return part, true
+		}
+	}
+
+	// If the peer is on a previous height that we have, help catch up.
+	blockStoreBase := blockStore.Base()
+	if blockStoreBase > 0 &&
+		0 < prs.Height && prs.Height < rs.Height &&
+		prs.Height >= blockStoreBase {
+		heightLogger := logger.With("height", prs.Height)
+
+		// if we never received the commit message from the peer, the block parts won't be initialized
+		if prs.ProposalBlockParts == nil {
+			blockMeta := blockStore.LoadBlockMeta(prs.Height)
+			if blockMeta == nil {
+				heightLogger.Error("Failed to load block meta",
+					"blockstoreBase", blockStoreBase, "blockstoreHeight", blockStore.Height())
+				return nil, false
+			}
+			ps.InitProposalBlockParts(blockMeta.BlockID.PartSetHeader)
+			// continue the loop since prs is a copy and not affected by this initialization
+			return nil, true // continue OUTER_LOOP
+		}
+		part := pickPartForCatchup(heightLogger, rs, prs, blockStore)
```

Comment:
In the original code, if we don't have the part to send, we: 
1. sleep: time.Sleep(conR.conS.config.PeerGossipSleepDuration)
2. continue OUTER_LOOP (we do that in any case, after calling gossipDataForCatchup)

If we have the part, we continue to OUTER_LOOP. If we don't, we don't continue.

Different behavior again.

Reply:
This is indeed a different behavior. As commented for line 727, if sending the `part` to the peer fails, the main routine should sleep. This is achieved with `continueLoop = false`.